[Kubernetes secret provider] Add cache for the secrets

constanca-m opened the pull request:
## What does this PR do?

Currently, we make a request to the API Server every time we need to get the value of a secret. This issue is better explained [here](https://github.com/elastic/elastic-agent/issues/3594). Discussion also present in the issue.

With this PR, we use a map to store the secrets like this:

1. Once the kubernetes secret provider starts running, we launch a go routine that will update the cache every TTL minutes. [TTL](https://github.com/constanca-m/elastic-agent/blob/dadc79b3d338bd2f4b37e3488c453d145450e436/internal/pkg/composable/providers/kubernetessecrets/config.go#L14) is a new option in the configuration of the provider. By default it is set to 10 minutes.
2. To update the cache, we range over every secret stored there, and make a new request to the API Server to obtain the current value.

The secrets are accessed from the outside through the function [Fetch](https://github.com/constanca-m/elastic-agent/blob/dadc79b3d338bd2f4b37e3488c453d145450e436/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go#L59). If we try to retrieve a secret that is not present in our map, then we make a request to the API Server to obtain its value and store it in the map. This function never causes an update on the secret values.

The only difference from this approach to the current one is that we know store the values of the secrets and secret values.

**Warning**:  Sometimes the secret values are not updated in time and we can be using incorrect ones until the update happens. This is also happening now.



## Why is it important?

To stop overwhelming the API Server with secret requests.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [ ] I have added an integration test or an E2E test


## How to test this PR locally

1. Clone this branch.
2. Follow [these steps](https://github.com/constanca-m/elastic-agent/blob/kubernetes-cache-secrets/README.md#testing-elastic-agent-on-kubernetes) from the README file.

## Related issues

- Relates to https://github.com/elastic/elastic-agent/issues/3594


## Screenshots

There should not be a change in behavior apart from a decrease in calls to the API Server. Everything else works as expected/before:

![image](https://github.com/elastic/elastic-agent/assets/113898685/526d9202-91d1-4875-89af-d0482e08b9aa)

elasticmachine commented:
Pinging @elastic/elastic-agent (Team:Elastic-Agent)

constanca-m commented on internal/pkg/composable/providers/kubernetessecrets/config.go:
```diff
@@ -10,4 +10,10 @@ import "github.com/elastic/elastic-agent-autodiscover/kubernetes"
+}
+
+func (c *Config) InitDefaults() {
+	c.TTL = 10
```
I honestly don't know what should be best. But we were updating on bigger intervals than 1. I can change to 1 as default, and then the user can update it as they wish.

gizas replied:
Can you please check with 1min to see if we have less api calls?  (now rememberd saying 3-5-7 minutes)
And we can decide then

constanca-m replied:
When I checked, we were having around 24 calls at the same time because we were getting calls from every resource. Now to obtain the same number of calls, even with TTL set to 1, we would need 24 minutes.

gizas commented on internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go:
```diff
@@ -46,19 +50,102 @@ func ContextProviderBuilder(logger *logger.Logger, c *config.Config, managed boo
+func (p *contextProviderK8sSecrets) updateCache() {
+	p.secretsCacheMx.Lock()
+	for name := range p.secretsCache {
+		newValue, ok := p.fetchSecret(name)
```
Shall we first check the value of each secret and if it is not the same then update it?

constanca-m replied:
I believe that getting or putting a new value into a map has the same complexity, so it does not make a difference. Would it make it more clear for readability? If so, I can add an `if` statement to check it before updating.

gizas replied:
Ok leave it. Ignore that comment

blakerouse commented on internal/pkg/composable/providers/kubernetessecrets/config.go:
```diff
@@ -10,4 +10,10 @@ import "github.com/elastic/elastic-agent-autodiscover/kubernetes"
 	KubeConfig        string                       `config:"kube_config"`
 	KubeClientOptions kubernetes.KubeClientOptions `config:"kube_client_options"`
+
+	TTL int `config:"ttl"`
```
You wrap this value with `time.Duration` when it is being used. Why not just use `time.Duration` in the configuration? That would allow the TTL to be set using `10s` instead of it being set to a specific unit.

constanca-m replied:
> Why not just use time.Duration in the configuration? 

Like it is now as a string? Or use time duration exactly in this place?

constanca-m replied:
I left it as a string, because I don't see the place to set it to the default in case the user sets it wrong :thinking:

blakerouse commented on internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go:
```diff
@@ -46,19 +50,102 @@ func ContextProviderBuilder(logger *logger.Logger, c *config.Config, managed boo
+
+// Update the secrets in the cache every TTL minutes
+func (p *contextProviderK8sSecrets) updateSecrets(ctx context.Context) {
+	d := time.Duration(p.config.TTL) * time.Minute
```
See my comment about about changing `p.config.TTL` to a `time.Duration` in the structure.

constanca-m replied:
Thanks, I updated the configuration according to this

blakerouse reviewed (changes requested):
The biggest issue I see in this PR is how does a key get removed from the cache if it is no longer referenced from the policy? Let say the policy references `kubernetes_secrets.somenamespace.somesecret.value1` and then changes to `kubernetes_secrets.somenamespace.somesecret.value2`. This change will keep caching `kubernetes_secrets.somenamespace.somesecret.value1` forever.

To solve the problem you should add a last accessed time per cached key, and during each update cycle and based on an interval you should remove keys that have not been accessed. That will solve the issue I described above.

constanca-m commented:
> > To solve the problem you should add a last accessed time per cached key
> 
> @blakerouse thanks for this . Makes sense.

And should we give that option also to the user or should we enforce that time ourselves @gizas ?

blakerouse commented on internal/pkg/composable/providers/kubernetessecrets/config.go:
```diff
@@ -10,4 +10,12 @@ import "github.com/elastic/elastic-agent-autodiscover/kubernetes"
 	KubeConfig        string                       `config:"kube_config"`
 	KubeClientOptions kubernetes.KubeClientOptions `config:"kube_client_options"`
+
+	TTL string `config:"ttl"`
```
Why are you making this a string? Make it a `time.Duration`. Elastic Agent uses https://github.com/elastic/go-ucfg which supports `time.Duration` as a native configuration type.

blakerouse commented on internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go:
```diff
@@ -46,19 +55,131 @@ func ContextProviderBuilder(logger *logger.Logger, c *config.Config, managed boo
+// Update the secrets in the cache every TTL minutes
+func (p *contextProviderK8sSecrets) updateSecrets(ctx context.Context) {
+	d, err := time.ParseDuration(p.config.TTL)
+	if err != nil {
```
You won't need to do this if you use a `time.Duration` in the structure. `go-ucfg` with Elastic Agent will handle this for you and report the errors when it parses the whole configuration.

constanca-m replied:
And if it fails, does it go back to the defaults?

blakerouse commented on internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go:
```diff
@@ -46,19 +55,131 @@ func ContextProviderBuilder(logger *logger.Logger, c *config.Config, managed boo
+			// if the secret has not been accessed in over 1h, delete it
+			// to avoid keeping secrets in cache that are no longer in use
+			diff := time.Since(data.lastAccess).Hours()
+			if diff > 1 {
```
If you make the TTL configurable, why not make this configuration as a `time.Duration`?

blakerouse commented:
This is also missing unit tests coverage for any of these additions, that needs to be added as well.

constanca-m commented:
> This is also missing unit tests coverage for any of these additions, that needs to be added as well.

I just added the unit test to check the results inside the cache @blakerouse

blakerouse commented on internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets_test.go:
```diff
@@ -138,3 +138,110 @@ func Test_K8sSecretsProvider_FetchWrongSecret(t *testing.T) {
+	require.NoError(t, err)
+
+	// wait for ttl update
+	<-time.After(ttlUpdate)
```
Relying on time here leads to random test failures because its possible that its executed out of order when the goroutines are scheduled. If your going to wait then you should at least double the wait time to ensure that it happens.

Another more full proof solution would be to use `assert.Eventually` after the wait to allow some chance that it has not be updated yet.

constanca-m replied:
I updated the code to use `assert.Eventually`.

blakerouse commented on internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets_test.go:
```diff
@@ -138,3 +138,110 @@ func Test_K8sSecretsProvider_FetchWrongSecret(t *testing.T) {
+
+	// After TTL delete, secret should no longer be found in cache since it was never
+	// fetched during that time
+	<-time.After(ttlDelete)
```
Same as above on this wait.

blakerouse commented on internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go:
```diff
@@ -46,19 +55,116 @@ func ContextProviderBuilder(logger *logger.Logger, c *config.Config, managed boo
+		if diff > p.config.TTLDelete {
+			delete(p.secretsCache, name)
+		} else {
+			newValue, ok := p.fetchSecret(name)
```
This is another issue that I am seeing. The lock is being held when making API calls to fetch the secret. What is the worry that this blocks for a long time holding the lock and then causing the rendering of the policy to be blocked?

Is it possible to be multiple key values in a single API call? It might be better to change this to first get all the keys that new values are needed. Then make a single API call to get all the values (if possible), and then update the cache with the values.

constanca-m replied:
> Is it possible to be multiple key values in a single API call? 

It is not possible. The only way to get multiple secrets at once is to use the list function instead of get [here](https://github.com/elastic/elastic-agent/blob/b3ba2192a841d4e353ac625620a88643b87144ef/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go#L202). However, the options to filter the secrets we return are [these](https://github.com/kubernetes/kubernetes/blob/ad9b60e2c9ddb21e8b00cabbe27e639638a0ea88/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L322), and as we can see, there are none to filter by the name. So in the end we would be retrieving all the secrets from a namespace when there is no need.

blakerouse replied:
Dang. Okay, I am still worried about it holding the lock the entire time it is refreshing all the secrets.

cmacknz replied:
Does `fetchSecrets` have a timeout? What if you never get a response? What if you do get a response, but it takes one hour?

Now that `cacheTmp` is being rebuilt is holding the lock for each `fetchSecret` call still necessary? Can `cacheTmp` be built and then assigned to `p.secretsCache` with only that one line holding the lock?

Since we are iterating over secretsCache that would likely require duplicating the keys first.

> However, the options to filter the secrets we return are these, and as we can see, there are none to filter by the name. So in the end we would be retrieving all the secrets from a namespace when there is no need.

What places more load on the k8s API? Returning a single response that is unnecessarily large or multiple small requests made rapidly one after another?

We should bias towards making sure the k8s api-server remains stable over optimizing the request size if possible.

constanca-m replied:
> Does fetchSecrets have a timeout? What if you never get a response? What if you do get a response, but it takes one hour?

I added a timeout just now, so we will never be waiting for 1h.

> Now that cacheTmp is being rebuilt is holding the lock for each fetchSecret call still necessary?

I changed the implementation to this way:
1. Created the temporary cache.
2. Apply read lock and iterate over the entries of the secrets cache. If entry remains, then write on temporary cache.
3. Now that the iteration is over, apply writing lock to secrets cache and make it the same as the temporary one.

Is this ok this way @cmacknz ?

constanca-m commented on internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go:
```diff
@@ -46,19 +55,116 @@ func ContextProviderBuilder(logger *logger.Logger, c *config.Config, managed boo
+func (p *contextProviderK8sSecrets) addToCache(key string) (secretsData, bool) {
+	value, ok := p.fetchSecret(key)
+	data := secretsData{
+		value: value,
```
It's on purpose. `addToCache` is only used inside `getFromCache`.

https://github.com/elastic/elastic-agent/blob/9d948597bbd3448e3ff66693cf2431a89d05bb5f/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go#L124

`getFromCache` is the only function that updates `lastAcess`.

blakerouse replied:
I see, thanks.

constanca-m commented on internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go:
```diff
@@ -33,6 +34,14 @@ type contextProviderK8sSecrets struct {
 	clientMx sync.Mutex
 	client   k8sclient.Interface
+
+	secretsCacheMx sync.Mutex
```
Yes, fixed.

constanca-m commented on internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go:
```diff
@@ -46,19 +55,116 @@ func ContextProviderBuilder(logger *logger.Logger, c *config.Config, managed boo
+	for name, data := range p.secretsCache {
+		diff := time.Since(data.lastAccess)
+		if diff > p.config.TTLDelete {
+			delete(p.secretsCache, name)
```
Fixed.

constanca-m commented:
Are there any blockers you still want to address @blakerouse , since you did not remove the requested changes?

blakerouse commented:
> Are there any blockers you still want to address @blakerouse , since you did not remove the requested changes?

Yes I am still waiting on my comment about holding the lock the entire time the code is refreshing the cache. I think that is still a big issue that needs to be looked at.

constanca-m commented on internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go:
```diff
@@ -46,38 +55,152 @@ func ContextProviderBuilder(logger *logger.Logger, c *config.Config, managed boo
+}
+
+// Update the secrets in the cache every TTL minutes
+func (p *contextProviderK8sSecrets) updateSecrets(ctx context.Context) {
+	timer := time.NewTimer(p.config.TTLUpdate)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-timer.C:
+			p.updateCache()
+			timer.Reset(p.config.TTLUpdate)
+		}
 	}
+}
```
> General question: why do we update all secrets periodically instead of adding/refreshing entries based on access?

We cannot do that, because otherwise we would need to make a request to the API Server every time we want to access the value. That is the [issue](https://github.com/elastic/elastic-agent/issues/3594) that led to this PR.

> If we set a TTL for deletion (basically the time we consider a cache value "not stale") and in case of cache miss we fetch the specific secret value woudn't that be more efficient ?

You mean to just update the secret after TTL delete has passed? TTL delete is quite a long duration.

> it's just to understand the rationale behind having a cache where (almost) all entries get refreshed at once: do we have an idea of what the access pattern of secrets would be ?

I am not sure, this a feature request. In the issue I [linked](https://github.com/elastic/elastic-agent/issues/3594) we discuss the possible workarounds. In the future, this cache will likely be replaced by a watcher, but we need to solve [this](https://github.com/elastic/beats/issues/34717) first.

constanca-m commented on internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go:
```diff
@@ -46,38 +55,152 @@ func ContextProviderBuilder(logger *logger.Logger, c *config.Config, managed boo
+	}
+
+	var pass string
+	p.secretsCacheMx.Lock()
+	data, ok := p.secretsCache[key]
+	data.lastAccess = time.Now()
+	pass = data.value
+	p.secretsCacheMx.Unlock()
```
> Are we using a write lock here because we are modifying the lastAccess attribute of a cache entry in place ?

Yes, we are updating the value there.

constanca-m commented on internal/pkg/composable/providers/kubernetessecrets/config.go:
```diff
@@ -4,10 +4,22 @@
 	KubeConfig        string                       `config:"kube_config"`
 	KubeClientOptions kubernetes.KubeClientOptions `config:"kube_client_options"`
+
+	TTLUpdate time.Duration `config:"ttl_update"`
```
Updated.

cmacknz commented on internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go:
```diff
@@ -46,67 +55,189 @@ func ContextProviderBuilder(logger *logger.Logger, c *config.Config, managed boo
+
+	var pass string
+	p.secretsCacheMx.Lock()
+	data, ok := p.secretsCache[key]
+	data.lastAccess = time.Now()
```
`data` is not a pointer so you are getting a copy of it, so the update to `lastAccess` does not change the `lastAccess` that was originally stored in the map.

You are missing test coverage for two consecutive `Fetch` calls of the same value to ensure that `lastAccess` updates in the expected way.

Here's a sample modification to your test showing the problem:

```diff
diff --git a/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets_test.go b/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets_test.go
index 3fea2dc455..a7348c40a9 100644
--- a/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets_test.go
+++ b/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets_test.go
@@ -212,12 +212,33 @@ func Test_K8sSecretsProvider_Check_TTL(t *testing.T) {
        val, found := fp.Fetch(key)
        assert.True(t, found)
        assert.Equal(t, val, pass)
+
        fp.secretsCacheMx.RLock()
        assert.Equal(t, len(fp.secretsCache), 1)
-       assert.NotNil(t, fp.secretsCache[key])
-       assert.NotZero(t, fp.secretsCache[key].lastAccess)
+       cachedVal1 := fp.secretsCache[key]
        fp.secretsCacheMx.RUnlock()

+       assert.NotNil(t, cachedVal1)
+       assert.NotZero(t, cachedVal1.lastAccess)
+
+       val, found = fp.Fetch(key)
+       assert.True(t, found)
+       assert.Equal(t, val, pass)
+
+       // Arbitrary delay to ensure that the value returned by time.Now() used when setting TTL is not
+       // the same between Fetch() calls.
+       time.Sleep(time.Millisecond)
+
+       fp.secretsCacheMx.RLock()
+       assert.Equal(t, len(fp.secretsCache), 1)
+       cachedVal2 := fp.secretsCache[key]
+       fp.secretsCacheMx.RUnlock()
+
+       assert.NotNil(t, cachedVal2)
+       t.Logf("Fetch 1 LastAccess: %s", cachedVal1.lastAccess)
+       t.Logf("Fetch 2 LastAccess: %s", cachedVal2.lastAccess)
+       assert.NotEqual(t, cachedVal1.lastAccess, cachedVal2.lastAccess)
+
        // Update the secret and check after TTL time, the secret value is correct
        newPass := "new-pass"
        secret = &v1.Secret{

```

```sh
❯ go test ./internal/pkg/composable/providers/kubernetessecrets/... -run Test_K8sSecretsProvider_Check_TTL

# github.com/elastic/elastic-agent/internal/pkg/composable/providers/kubernetessecrets.test
ld: warning: ignoring duplicate libraries: '-lproc'
--- FAIL: Test_K8sSecretsProvider_Check_TTL (2.22s)
    kubernetes_secrets_test.go:238: Fetch 1 LastAccess: 2023-12-05 16:35:13.749574 -0500 EST m=+0.016378918
    kubernetes_secrets_test.go:239: Fetch 2 LastAccess: 2023-12-05 16:35:13.749574 -0500 EST m=+0.016378918
    kubernetes_secrets_test.go:240:
                Error Trace:    /Users/cmackenzie/go/src/github.com/elastic/elastic-agent/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets_test.go:240
                Error:          Should not be: time.Date(2023, time.December, 5, 16, 35, 13, 749574000, time.Local)
                Test:           Test_K8sSecretsProvider_Check_TTL
FAIL
FAIL    github.com/elastic/elastic-agent/internal/pkg/composable/providers/kubernetessecrets    2.754s
FAIL
```

constanca-m replied:
Thanks @cmacknz for such a complete example.

I checked this:

> data is not a pointer so you are getting a copy of it, so the update to lastAccess does not change the lastAccess that was originally stored in the map.

And I ran the test. I added a println on the code right where the last access changes:

```
=== RUN   Test_K8sSecretsProvider_Check_TTL
2023-12-06 08:26:10.267182544 +0100 CET m=+0.016935938
2023-12-06 08:26:10.267266524 +0100 CET m=+0.017019919
    kubernetes_secrets_test.go:239: Fetch 1 LastAccess: 2023-12-06 08:26:10.267266524 +0100 CET m=+0.017019919
    kubernetes_secrets_test.go:240: Fetch 2 LastAccess: 2023-12-06 08:26:10.267266524 +0100 CET m=+0.017019919
    kubernetes_secrets_test.go:241: 
        	Error Trace:	/home/c/Documents/Git-Code/Elastic-Forks/elastic-agent/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets_test.go:241
        	Error:      	Should not be: time.Date(2023, time.December, 6, 8, 26, 10, 267266524, time.Local)
        	Test:       	Test_K8sSecretsProvider_Check_TTL
--- FAIL: Test_K8sSecretsProvider_Check_TTL (0.01s)
```

So, as you can see, the last value got updated. But since `cachedVal1` and `cachedVal2` are the same, this assertion will always fail `assert.NotEqual(t, cachedVal1.lastAccess, cachedVal2.lastAccess)`. Notice that the map contains the exact values, not copies:

```
type contextProviderK8sSecrets struct {
        ...
	secretsCache   map[string]*secretsData <---------------------- pointer here
}
```

cmacknz replied:
Yes you are correct, my example proves myself wrong I just needed to read it at the start of my day instead of the end :)

cmacknz commented on internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go:
```diff
@@ -46,67 +55,189 @@ func ContextProviderBuilder(logger *logger.Logger, c *config.Config, managed boo
+
+	// if value is still not present in cache, it is possible we haven't tried to fetch it yet
+	if !ok {
+		data, ok := p.addToCache(key)
```
It is possible that another goroutine acquires could call addToCache before this goroutine. That is this goroutine is preempted immediately after `RUnlock`, then another goroutine calls `addToCache`, and `addToCache` is then called again unnecessarily.

I think this is more of an optimization than a bug, since all it results in is an unnecessary call to the secrets API.

constanca-m replied:
I switched the reading lock to a writing lock, so this doesn't happen anymore.

cmacknz commented on internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go:
```diff
@@ -46,67 +55,189 @@ func ContextProviderBuilder(logger *logger.Logger, c *config.Config, managed boo
+
+		}
+	}
+	p.secretsCacheMx.RUnlock()
```
If this goroutine is preempted immediately after the `RUnlock` call here, and then another goroutine calls `addToCache` via a call to `Fetch` for a secret that wasn't previously in the map, then the write is lost and `p.secretsCache = cacheTmp` discards that secret.

You would need to iterate over `p.secretsCache` again while holding the write lock to check for values that were added when you released the `RLock` above to include those as well.

In theory the next call to `Fetch` would populate the missing value but you'd need to make a duplicate call to the k8s API for this.

cmacknz reviewed (commented):
Took another look, there is still at least one bug and a few cases I could spot where splitting the cache changes across separate "read lock" and "write lock" sections introduces concurrency problems. 

This is because the manipulation you are doing to the cache is no longer atomic across a single function body allowing the cache to change in the middle of the function execution when you unlock the read lock and acquire the write lock.

cmacknz commented on internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go:
```diff
@@ -46,67 +55,183 @@ func ContextProviderBuilder(logger *logger.Logger, c *config.Config, managed boo
+	defer cancel()
+
+	resultCh := make(chan Result, 1)
+	go p.fetchSecret(ctxTimeout, key, resultCh)
```
Why does this need to be in a goroutine if we are just going to synchronously wait for it to complete anyway?

constanca-m replied:
You are right. I will remove the `go`

cmacknz reviewed (commented):
One last thought, it seems like it would be straightforward to add a configuration entry that disables the cache.

That way if there is some unexpected surprise we can revert to the previous behavior without needing a release.

constanca-m commented:
>  it seems like it would be straightforward to add a configuration entry that disables the cache.

I added this option in my last commit. I also added the proper unit test for it.

constanca-m commented on internal/pkg/composable/providers/kubernetessecrets/config.go:
```diff
@@ -4,10 +4,26 @@
+	RefreshInterval time.Duration `config:"refresh_interval"`
+	TTLDelete       time.Duration `config:"ttl_delete"`
+	RequestTimeout  time.Duration `config:"request_timeout"`
+	DisableCache    bool          `config:"disable_cache"`
```
I will add the prefix `cache_` to all of those then, and this way we will be following the same logic as we followed for kubernetes leader:
https://github.com/elastic/elastic-agent/blob/e3c0695757b3799335214ae20af2e847c0d8cc00/internal/pkg/composable/providers/kubernetesleaderelection/config.go#L14-L20

gizas commented on internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go:
```diff
@@ -46,67 +55,201 @@ func ContextProviderBuilder(logger *logger.Logger, c *config.Config, managed boo
+		}
+	}
+	p.secretsCache = cacheTmp
+	p.secretsCacheMx.Unlock()
```
@blakerouse I can understand your concern. For now the option to disable the cache has been added [0f8d86a](https://github.com/elastic/elastic-agent/pull/3822/commits/0f8d86a4955f6737e74b504e83f11a6188fcb3fb)

So in case of any problem like you describe we can revert to ption where we just get the secrret [here](https://github.com/elastic/elastic-agent/blob/0f8d86a4955f6737e74b504e83f11a6188fcb3fb/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go#L240) with a small lock to get the [client](https://github.com/elastic/elastic-agent/blob/0f8d86a4955f6737e74b504e83f11a6188fcb3fb/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go#L223)

Can we agree that Syncronisation of cache can be  done in a new story? We also investigate this on other cases for kubernetes watchers and if we have any update we can link this issue as well. 
For now I will approve this story

MichaelKatsoulis replied:
I agree that holding the lock during the whole duration of update is not a good solution due to the blocking it can create to Fetch. I understand from the code that the reason to lock, is because we read all the keys and the lastaccess from the cache in the iteration. We actually write to the cache only after all secrets are updated. Why don't we keep the keys of the cache map in a slice and only lock the cache when we actually read the last access value for each key. Something like this 

```
func (p *contextProviderK8sSecrets) updateCache() {
	cacheTmp := make(map[string]*secretsData)

	p.secretsCacheMx.Lock()
	secretKeys := make([]string, len(p.secretsCache))
	i := 0
	for k := range p.secretsCache {
		secretKeys[i] = k
		i++
	}
	p.secretsCacheMx.Unlock()
	for _, name := range secretKeys {
		p.secretsCacheMx.Lock()
		lastAccess := p.secretsCache[name].lastAccess
		p.secretsCacheMx.Unlock()
		diff := time.Since(lastAccess)

		if diff < p.config.TTLDelete {
			value, ok := p.fetchSecretWithTimeout(name)
			if ok {
				newData := &secretsData{
					value:      value,
					lastAccess: lastAccess,
				}
				cacheTmp[name] = newData
			}

		}
	}
	p.secretsCacheMx.Lock()
	p.secretsCache = cacheTmp
	p.secretsCacheMx.Unlock()
}
```

constanca-m replied:
That has already been suggested, and @cmacknz explains in this https://github.com/elastic/elastic-agent/pull/3822#discussion_r1416314549 why that does not work. @MichaelKatsoulis

MichaelKatsoulis replied:
You mean that they keys can change by fetch method which can create a new entry in the cache, and then in update due to the temp cache we will override and lose the new key? 
Because if this is the problem, it could be mitigated if we then merge the `p.secretsCache` and `cacheTmp` while holding the cache.

constanca-m replied:
> You mean that they keys can change by fetch method which can create a new entry in the cache, and then in update due to the temp cache we will override and lose the new key?

Yes.

> Because if this is the problem, it could be mitigated if we then merge the p.secretsCache and cacheTmp while holding the cache.

We cannot do that. It is explained in this https://github.com/elastic/elastic-agent/pull/3822#discussion_r1406557666 why not

MichaelKatsoulis replied:
In that comment it is explained that deleting a key from a map does not delete it from memory. I am saying that after the update is done and new keys are added in the temp cache we can do a merge of the temp cache and p.secretsCache and keep the keys that may have been added while we were updating. See this example:

```
p.secretsCacheMx.Lock()
newCache := mergeMaps(p.secretsCache, cacheTmp, p.config.TTLDelete)
p.secretsCache = newCache
p.secretsCacheMx.Unlock()
```

```
func mergeMaps[K comparable, V any](m1 map[K]V, m2 map[K]V, ttl time.Duration) map[K]V {
	merged := make(map[K]V)
	for key, value := range m1 {
		lastAccess := any(value).(secretsData).lastAccess
		diff := time.Since(lastAccess)
		if diff < ttl {
			merged[key] = value
		}
	}
	for key, value := range m2 {
		merged[key] = value
	}
	return merged
}
```

blakerouse replied:
I agree with @MichaelKatsoulis suggestion but I believe you need to make an approvement to the mergeMaps. I believe that it is updating the lastAccess time from the `m2` map. It should most likely only overwrite the value and not the `lastAccess` timestamp.

MichaelKatsoulis replied:
Yeah, in case there was a key accessed by fetch, while the update took place, then the last access of that key will have been updated. And by 
```
for key, value := range m2 {
		merged[key] = value
	}
```
the new last access will be overridden by lets say the old last access of the cache. 
Maybe using generics in this use case does not add anything and makes it more complicated as we already know the data types of the maps.
Something like this could work
```
func mergeMaps(m1, m2 map[string]*secretsData, ttl time.Duration) map[string]*secretsData {
	merged := make(map[string]*secretsData)
	for key, value := range m1 {
		lastAccess := value.lastAccess
		diff := time.Since(lastAccess)
		if diff < ttl {
			merged[key] = value
		}
	}
	for key, value := range m2 {
		if val, ok := merged[key]; ok {
			merged[key] = &secretsData{
				value:      value.value,
				lastAccess: val.lastAccess,
			}
		} else {
			merged[key] = value
		}
	}
	return merged
}
```

blakerouse replied:
I don't believe you want that last else statement. If the key was removed due to ttl in the previous loop you don't want to add it back in the case that it was fetched in the cache refresh, otherwise I believe the function will work and prevent the holding of the lock for the full refresh.

constanca-m replied:
I made a new commit, so now:

1. We hold only a read lock to update the secrets.
2. After that, we merge the maps.

Is it ok this way @blakerouse @MichaelKatsoulis ?

MichaelKatsoulis replied:
@blakerouse you are right. I somehow thought that update adds extra keys as well but that is not the case.

constanca-m replied:
It is a reading lock now @blakerouse (this file preview is not updated to the last commit)

constanca-m replied:
I updated the code to do that in the latest commit @blakerouse

MichaelKatsoulis commented on internal/pkg/composable/providers/kubernetessecrets/config.go:
```diff
@@ -4,10 +4,26 @@
+	RefreshInterval time.Duration `config:"cache_refresh_interval"`
+	TTLDelete       time.Duration `config:"cache_ttl_delete"`
+	RequestTimeout  time.Duration `config:"cache_request_timeout"`
+	DisableCache    bool          `config:"cache_disable_cache"`
```
```suggestion
	DisableCache    bool          `config:"cache_disable"`
```

MichaelKatsoulis commented on internal/pkg/composable/providers/kubernetessecrets/config.go:
```diff
@@ -4,10 +4,26 @@
 	KubeClientOptions kubernetes.KubeClientOptions `config:"kube_client_options"`
+
+	RefreshInterval time.Duration `config:"cache_refresh_interval"`
+	TTLDelete       time.Duration `config:"cache_ttl_delete"`
```
```suggestion
	TTLDelete       time.Duration `config:"cache_ttl"`
```

MichaelKatsoulis commented on internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go:
```diff
@@ -46,67 +55,223 @@ func ContextProviderBuilder(logger *logger.Logger, c *config.Config, managed boo
+		}
+	}
+	for name, data := range updatedMap {
+		merged[name] = data
```
In https://github.com/elastic/elastic-agent/pull/3822/files#r1425592648 I explain why this will override the last access of the secrets that were accessed by fetch , while the update was taking place. 
The right thing to do here would be 
```
for name, data := range updatedMap {
		if val, ok := merged[name]; ok {
			merged[name] = &secretsData{value: data.value, lastAccess: val.lastAccess}
		}
	}
```

constanca-m replied:
Thanks, I updated the code to fix the mistake.

MichaelKatsoulis commented on internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go:
```diff
@@ -46,67 +55,223 @@ func ContextProviderBuilder(logger *logger.Logger, c *config.Config, managed boo
+}
+
+func (p *contextProviderK8sSecrets) getFromCache(key string) (string, bool) {
+	p.secretsCacheMx.Lock()
```
I believe now RLock and RUnlock should be used.

constanca-m replied:
Not possible, because we are calling `addToCache` in this function, and we need that lock

MichaelKatsoulis replied:
TBH I don't understand why we need to hold the lock while we are waiting to fetchSecretWithTimeout

constanca-m replied:
We are adding a value to the cache, and then we are returning the value that is inside the cache. If we don't hold the lock there, we get a racing condition from the tests.

MichaelKatsoulis replied:
Ok can you explain to me what is wrong with this code? I am not holding the lock while waiting for 
fetchSecretWithTimeout. 

```
func (p *contextProviderK8sSecrets) getFromCache(key string) (string, bool) {
	p.secretsCacheMx.RLock()
	defer p.secretsCacheMx.RUnlock()

	data, ok := p.secretsCache[key]
	if ok {
		data.lastAccess = time.Now()
		pass := data.value
		return pass, ok
	}
	data, ok = p.addToCache(key)
	// if it was not possible to fetch the secret, return
	return data.value, ok
}

func (p *contextProviderK8sSecrets) addToCache(key string) (*secretsData, bool) {
	valid := p.validateKey(key)
	if !valid {
		return &secretsData{
			value: "",
		}, false
	}

	value, ok := p.fetchSecretWithTimeout(key)
	data := &secretsData{
		value: value,
	}
	if ok {
		p.secretsCacheMx.Lock()
		p.secretsCache[key] = data
		p.secretsCacheMx.Unlock()
	}
	return data, ok
}
```

MichaelKatsoulis replied:
Yes because of the defer wich will keep the read lock while adToCache runs. Actually what it needs is to just remove the read lock after reading

```
func (p *contextProviderK8sSecrets) getFromCache(key string) (string, bool) {
	p.secretsCacheMx.RLock()
	data, ok := p.secretsCache[key]
	p.secretsCacheMx.RUnlock()
	if ok {
		data.lastAccess = time.Now()
		pass := data.value
		return pass, ok
	}
	data, ok = p.addToCache(key)
	return data.value, ok
}
```
One of the tests is not passing though ...

MichaelKatsoulis replied:
From my side it's good to go! Thanks for patience!

blakerouse commented on internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go:
```diff
@@ -46,67 +55,229 @@ func ContextProviderBuilder(logger *logger.Logger, c *config.Config, managed boo
+
+	// if value is still not present in cache, it is possible we haven't tried to fetch it yet
+	if !ok {
+		value, ok := p.addToCache(key)
```
Why grab the lock inside of the function to set the new value. Then only 4 lines below grab it again to set the lastAccess time?

Why not just set `lastAccess` inside of `addToCache` and then return inside of the if statement? That would prevent the need to grab the lock twice for the same key.

constanca-m replied:
> Why not just set lastAccess inside of addToCache and then return inside of the if statement?

Because then we would be changing `lastAccess` in two functions, when we don't need it. The logic is that `lastAccess` is only changed when it is required (by calling the get from cache).

blakerouse replied:
To me its worse to grab a lock, release the lock, grab the lock again, then release it again all in the same path. It should grab the lock do what it needs to do and release it.

blakerouse reviewed (approved):
Thanks for all the changes and fixes. I am glad a solution was able to be created for not holding the lock the entire time a refresh loop occurs.

My last point of hold/releasing followed by another hold/release is not a blocker, just seems like it would be less CPU cycles to grab the lock once. I understand your trying to be my DRY with the code, but for something as simple as setting a single attribute timestamp I think less lock contention would be worth it.
